Refactor import best practices (#17473)
cockroachdb/docs: src/current/v22.1/import-performance-best-practices.md

docs_area: migrate
---

This page provides best practices for optimizing [import](import-into.html) performance in CockroachDB.

`IMPORT INTO` is the fastest method to ingest data into CockroachDB but it requires taking the target table offline for the duration of the import. `IMPORT INTO` is a good choice for initial data migrations and data migrations that can tolerate table downtime. If you cannot tolerate table unavailability, we recommend using [`COPY FROM`](copy-from.html) instead.

> Import performance primarily depends on the amount of data that you want to import. However, there are three actions you can take before importing that have a significant impact on the amount of time it will take to run an import:

Reviewer: ```suggestion
Import performance primarily depends on the amount of data that you want to import. However, there are three actions you should take before importing that will have a significant impact on the amount of time it takes to run an import:
```
Minor wording suggestion to make this a stronger suggestion. Feel free to ignore

- [Choose a performant import format](#choose-a-performant-import-format)
- [Split your data into multiple files](#split-your-data-into-multiple-files)
- [Sort your data](#sort-your-data)

{{site.data.alerts.callout_info}}
If the import size is less than 100 GiB, then you do not need to do anything to optimize performance. For such small datasets, the import should run quickly, regardless of the settings.
{{site.data.alerts.end}}

## Choose a performant import format

Different import file formats do not have the same performance due to the way they are processed by CockroachDB. The fastest import file formats are:

1. [`CSV`](migrate-from-csv.html) or [`DELIMITED DATA`](import-into.html) (both have about the same import performance).
1. [`AVRO`](migrate-from-avro.html).

We recommend formatting your import files as `CSV` or `AVRO`. These formats can be processed in parallel by all the nodes in the cluster, which increases performance. To import in these formats, use [`IMPORT INTO`](import-into.html).

{% include {{ page.version.version }}/import-table-deprecate.md %}

### Import the schema separately from the data

> When importing into a new table split your dump data into two files:

Reviewer: ```suggestion
When importing into a new table, split your dump data into two files:
```

1. A SQL file containing the table schema.
1. A CSV, delimited, or AVRO file containing the table data.

Convert the schema-only file using the [Schema Conversion Tool](../cockroachcloud/migrations-page.html). The Schema Conversion Tool automatically creates a new {{ site.data.products.serverless }} database with the converted schema. {% include cockroachcloud/migration/sct-self-hosted.md %}

Then use the [`IMPORT INTO`](import-into.html) statement to import the CSV data into the newly created table:

~~~ sql
> IMPORT INTO customers (id, name)
CSV DATA (
    'https://s3-us-west-1.amazonaws.com/cockroachdb-movr/datasets/employees-db/pg_dump/customers.csv'
);
~~~

This method has the added benefit of alerting on potential issues with the import sooner; that is, you will not have to wait for the file to load both the schema and data just to find an error in the schema.

### Import into a schema with secondary indexes

> When importing data into a table with secondary indexes, the import job will ingest the table data and required secondary index data concurrently. This may result in a longer import time compared to a table without secondary indexes. However, this typically adds less time to the initial import than following it with a separate pass to add the indexes. As a result, importing tables with their secondary indexes is the default workflow, an effective strategy for most migrations.

Reviewer: ```suggestion
When importing data into a table with [secondary indexes](schema-design-indexes.html), the import job will ingest the table data and required secondary index data concurrently. This may result in a longer import time compared to a table without secondary indexes. However, this typically adds less time to the initial import than following it with a separate pass to add the indexes. As a result, importing tables with their secondary indexes is the default workflow, an effective strategy for most migrations.
```

However, in **large** imports (that is, datasets larger than 100 GiB in total size), it may be preferable to temporarily remove the secondary indexes from the schema, perform the import, and then re-create the indexes separately. Write operations on tables with many secondary indexes take longer to complete, and importing large datasets have a greater risk of timeouts. Removing the table's secondary indexes allows you to separate the initial data import from the secondary index creation operation, and the total import time is lower. Separating these operations also provides increased visibility into each operation's progress, and the ability to retry each operation independently if you encounter errors or timeouts.

When recreating the secondary indexes, execute the [`CREATE INDEX`](create-index.html) statements one at a time for best performance. For example:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE INDEX idx1 ON table1;
CREATE INDEX idx2 ON table1;
CREATE INDEX idx3 ON table1;
~~~

> - [Remove the secondary indexes](drop-index.html)

Reviewer: Should these three bullets' links just be integrated into the "However, in large imports..." paragraph? 

If you want to keep these as bullets, I think they would benefit from a intro sentence before them.

- [Perform the import](import-into.html)
- [Create a secondary index](schema-design-indexes.html#create-a-secondary-index)

### Temporarily remove foreign keys

> When importing data into a table with foreign keys, temporarily remove the foreign keys, and add them after the initial migration with an [`ALTER TABLE` statement](add-constraint.html#add-the-foreign-key-constraint-with-cascade).

Reviewer: ```suggestion
When importing data into a table with [foreign keys](foreign-key.html), temporarily remove the foreign keys, and add them after the initial migration with an [`ALTER TABLE` statement](add-constraint.html#add-the-foreign-key-constraint-with-cascade).
```

Cockroach Labs recommends keeping the files to a maximum file size of 4 GB unless the files are streamed (for example, [cloud storage locations](use-cloud-storage-for-bulk-operations.html) will stream the data to CockroachDB), and to keep each file size similar across the dataset. For example, if you are importing a 9 GB dataset that was split into 3 files into a 3 node cluster, keep each file around 3 GB in size if possible. Don't split the data into two 4 GB files, and one 1 GB file.
{{site.data.alerts.end}}

For maximum performance each split file should be sorted within and across all files, meaning that if you were to sort the rows in each file there would be no overlapping data in any other file. For example, suppose your table has an alphabetic string primary key and you were importing the data into a 3 node cluster.

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE TABLE contacts (email STRING PRIMARY KEY, first_name TEXT, last_name TEXT);
~~~

> You should split the files so that the first file contains rows where the primary key begins with the letters "a" to "i," the second file "j" to "r," and the third file "s" to "z". No file contains duplicate rows, and within each file the data is sorted alphabetically by the primary key.

Reviewer: ```suggestion
You should split the files so that the first file contains rows where the [primary key](primary-key.html) begins with the letters "a" to "i," the second file "j" to "r," and the third file "s" to "z". No file should contain duplicate rows, and within each file the data is sorted alphabetically by the primary key.
```
